app/src/test/java/app/envelop/ui/login/LoginViewModelTest.kt

```diff
@@ -0,0 +1,78 @@
+package app.envelop.ui.login
+
+import app.envelop.common.Operation
+import app.envelop.domain.LoginService
+import app.envelop.ui.common.Finish
+import app.envelop.ui.common.LoadingState
+import com.nhaarman.mockitokotlin2.any
+import com.nhaarman.mockitokotlin2.doReturn
+import com.nhaarman.mockitokotlin2.mock
+import com.nhaarman.mockitokotlin2.whenever
+import io.reactivex.Single
+import org.junit.Assert.assertEquals
+import org.junit.Before
+import org.junit.Test
+
+
+class LoginViewModelTest {
+
+    val loginService = mock<LoginService>()
+    val loginViewModel = LoginViewModel()
+
+    @Before
+    fun setUp() {
+        loginViewModel.loginService = loginService
+    }
+
+    @Test
+    fun isLoggingIn() {
+        whenever(loginService.login()).doReturn(Single.just(Operation()))
+
+        val valueStream = loginViewModel.isLoggingIn().test()
+        loginViewModel.loginClick()
+
+        assertEquals(
+            valueStream.values().last(),
+            LoadingState.Idle
+        )
+    }
+
+    @Test
+    fun errorsLogin() {
+        whenever(loginService.login()).doReturn(Single.just(Operation(throwable = RuntimeException())))
+
+        val errorStream = loginViewModel.errors().test()
+        loginViewModel.loginClick()
+
+        errorStream.assertValue(LoginViewModel.Error.LoginError)
+    }
+
+    @Test
+    fun errorsAuth() {
+        whenever(loginService.finishLogin(any())).doReturn(
+            Single.just(
+                Operation(
+                    throwable = LoginService.UsernameMissing(
+                        ""
+                    )
+                )
+            )
+        )
+
+        val errorStream = loginViewModel.errors().test()
+        loginViewModel.authDataReceived("")
+
+        errorStream.assertValue(LoginViewModel.Error.UsernameMissing)
+    }
+
+    @Test
+    fun finishToMain() {
+        whenever(loginService.finishLogin(any())).doReturn(Single.just(Operation()))
+
+        val finishStream = loginViewModel.finishToMain().test()
+        loginViewModel.authDataReceived("")
+
+        val finishResult = finishStream.values().first()
```

Comment:
`finishStream.assertValue(Finish.Result.Ok)` here would also work.

Reply:
I tried, but it seems it can't assert the type of value...
![Screenshot 2020-09-08 at 10 30 26](https://user-images.githubusercontent.com/20781703/92459219-4f9d2380-f1be-11ea-95c3-a5564b166c1c.png)